### PR TITLE
chain: use batch rpc client in mempool poller

### PR DESCRIPTION
**Pull request by yyforyongyu**
This PR replaces the normal client with a batch client created from `rpcclient.NewBatch`. All the `getrawtransaction` requests are now batched with a default size of 10,000, and the batch will wait for 1 second before attempting the next one.

This change is made so less pressure is placed on bitcoind as when the mempool is full, there will be loads of `getrawtransaction` calls made, resulting in a slow response for other RPC calls such as `getblockchaininfo`, causing issues in other systems.

The mod files are also updated and a similar check is added here(copied from `lnd`)

### CODE REVIEW

File: wallet/txauthor/go.mod
Comment (guggero) [inline wallet/txauthor/go.mod:10]: Reminder to push new tags for all those submodules after merging the PR.

**Review by guggero — COMMENTED** [review]
Did a first pass, looks really good! Hope this'll quite significantly reduce the load on `bitcoind`.

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go]: The RPC requests are already processed sequentially in `btcd` tho, so it won't have any effect here on that front. Unless we are concerning about the overhead of creating and destroying goroutines, I think it's better to let the runtime scheduler to handle them. Also `GOMAXPROCS` is already default to `runtime.NumCPU` 

> it probably still makes sense to limit the number of goroutines to prevent clogging on slower systems.

Have we experienced this before?

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go]: fixed

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go]: yeah I like it, added the change in a new commit to log the txid.

File: chain/mempool_test.go
Comment (yyforyongyu) [reply chain/mempool_test.go]: yep, patched the unit tests.

File: chain/mempool.go
Comment (guggero) [reply chain/mempool.go]: > Have we experienced this before?

Yes, at least I think that in certain situations (e.g. during itests) when we ran too many things in parallel, some goroutines would just be starved (didn't get any CPU time). Not sure if that gets worse with the number of goroutines or just with the number of processes running on a system.

Comment (guggero) [reply chain/mempool.go]: Yes. So when the CPU is already struggling, goroutines can get starved (at least it looked like it). But I'm not sure if that gets worse with the number of goroutines there are in total or whether that doesn't have any direct impact. Just some thoughts/observations to possibly help the decision here.

File: chain/mempool_test.go
Comment (guggero) [inline chain/mempool_test.go]: Looks like this has been addressed by the last commit.

File: wallet/txauthor/go.sum
Comment (guggero) [inline wallet/txauthor/go.sum:25]: What version of Go did you run this on? I'm getting a diff in `wallet/{txauthor,txrules,txsizes}` with Go 1.18 or later.

**Review by guggero — COMMENTED** [review]
Very nice, looks very good!
Just a bunch of questions/comments.

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go:175]: Cool updated.

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go:415]: yeah exactly, the upper system should also be shutting down atm.

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go]: What overhead are you referring to? I think under the hood `GetRawMempool` behaves the same as it calls `GetRawMempoolAsync` in `btcd`. Also I tried to use `GetRawMempool` but it doesn't work for batch client.

File: wallet/txauthor/go.sum
Comment (yyforyongyu) [reply wallet/txauthor/go.sum:25]: mine is `go version go1.20.5 darwin/arm64`. What are the diffs here? I think maybe it's due to my local go work setup, that it also makes sure this repo is in sync with `lnd`.

File: chain/mempool.go
Comment (guggero) [reply chain/mempool.go:411]: I guess it's just visually more apparent that there is _no_ pre-allocation/initialization. It's easier to miss the `0` IMO and the `make()` would subconsciously register as "oh, we're allocating something here" when scanning the code quickly.
But it's really just nit-picking at this point. But I guess it's nice if all code uses the same pattern.

File: chain/mempool.go
Comment (guggero) [reply chain/mempool.go]: Just a suspicion, but since I don't really know how the async RPC works under the hood I can't go into more detail. And if there is no `GetRawMempool` call on batch clients, then I guess that settles it anyway.

File: wallet/txauthor/go.sum
Comment (guggero) [reply wallet/txauthor/go.sum:25]: Here's the diff I get when running `go mod tidy`:
[diff.txt](https://github.com/btcsuite/btcwallet/files/12730000/diff.txt)

**Review by guggero — APPROVED** [review]
LGTM, pending resolution of the `go mod tidy` diff.

File: wallet/txauthor/go.sum
Comment (yyforyongyu) [reply wallet/txauthor/go.sum:25]: yeah very weird, for instance this line
```
github.com/golang/protobuf v1.5.2/go.mod h1:XVQd3VNwM+JqD3oG2Ue2ip4fOMUkwXdXDdiuN0vRsmY=
```

looks like this is in sync with lnd's mod which also has this version.

**Comment by Roasbeef** [comment]
cc @ellemouton

**Comment by saubyk** [comment]
cc @Crypt-iQ for review. thanks

**Comment by Crypt-iQ** [comment]
Needs a rebase since the gettxspendingprevout change landed

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go:411]: cool yeah let's stick to this pattern then

Comment (guggero) [reply chain/mempool.go]: Ah, also, I guess when it comes to unit tests and deep equality tests, there is a difference between `var newTxids []*chainhash.Hash` and `newTxids := make([]*chainhash.Hash, 0)`: The first will give you a `nil` slice, while the second will give you an empty `[]*chainhash.Hash{}` slice. So if that was a "decode" function, we'd get a diff on such a field from before encoding and after, if there were no elements present.

Comment (yyforyongyu) [reply chain/mempool.go]: hmm so I had to revert the change to fix the linter,
```
Error: chain/mempool.go:411:2: Consider preallocating `newTxids` (prealloc)
	var newTxids []*chainhash.Hash
```

File: chain/bitcoind_rpc_events.go
Comment (Crypt-iQ) [inline chain/bitcoind_rpc_events.go:110]: The `getRawTxBatchSize` & `batchWaitInterval` members of `mCfg` don't get set if a non-default value is used

File: chain/bitcoind_zmq_events.go
Comment (Crypt-iQ) [inline chain/bitcoind_zmq_events.go:156]: Same as in the RPC case

File: chain/bitcoind_events.go
Comment (yyforyongyu) [reply chain/bitcoind_events.go:40]: Done

File: chain/bitcoind_zmq_events.go
Comment (yyforyongyu) [reply chain/bitcoind_zmq_events.go:138]: added back

File: chain/mempool.go
Comment (yyforyongyu) [reply chain/mempool.go]: Decreased by 10x. Not sure if the 32MB body size limit also applies for batch rpc calls? 

> Additionally, I think this might cause other goroutines calling bitcoind to have to wait a long time for the batch to finish

Probably. I assume the lookup will be fast in `bitcoind`, sth like `log(N)` if indexed. And because we are sending 10000x fewer requests here, it will actually make `bitcoind` faster to respond because of the overhead saved.

File: chain/bitcoind_zmq_events.go
Comment (yyforyongyu) [reply chain/bitcoind_zmq_events.go:158]: fixed

File: chain/mempool.go
Comment (Crypt-iQ) [reply chain/mempool.go]: I think any gain would come from less HTTP headers. `bitcoind` will still call the [same function N times](https://github.com/bitcoin/bitcoin/blob/744157ef1a0b61ceb714cc27c9ae158907aecdc9/src/rpc/rawtransaction.cpp#L261) and still acquire the main `cs_main` lock N times. I'd be curious to benchmark this in a reproducible way and see where any gain comes from, but not sure how feasible that is.

The 32MB body size applies for all incoming HTTP requests to bitcoind, it's set here: https://github.com/bitcoin/bitcoin/blob/744157ef1a0b61ceb714cc27c9ae158907aecdc9/src/httpserver.cpp#L457

File: chain/bitcoind_events.go
Comment (Crypt-iQ) [inline chain/bitcoind_events.go:40]: nit: I think `client` & `bClient` should be defined with their interfaces rather than both being `*rpcclient.Client` which is a little confusing. Alternatively, `bClient batchClient` works.

File: chain/bitcoind_rpc_events.go
Comment (Crypt-iQ) [inline chain/bitcoind_rpc_events.go:80]: nit: same thing with the type of `bClient`

**Review by Crypt-iQ — APPROVED** [review]
LGTM besides nit

File: chain/bitcoind_events.go
Comment (yyforyongyu) [reply chain/bitcoind_events.go]: updated the method signature - yeah I also think it's a bit confusing and prone to errors, and think this needs to be changed in `btcd` so we can have a `rpcclient.BatchClient`.

**Review by Crypt-iQ — APPROVED** [review]

**Pull request merged**
